### PR TITLE
[TTS] Implement new TextToSpeech dataset

### PR DESCRIPTION
**Pull request by rlangman**
# What does this PR do ?

Add initial implementation of new TTS data loader which supports multi-manifest training and uses mostly preprocessed features.

The data loader is intended to be for TTS acoustic model and as such includes some specific functionality for text processing. Other models like vocoder will need another data loader, with most of the common code being shared in the features & tts_dataset_utils.

**Collection**: [TTS]

# Changelog 
- Implement TextToSpeechDataset class (following ASR naming convention of having separate X to X datasets for each task). 
- Add example FastPitch config for using new dataset.
- Modify feature.yaml files to use relative paths so they can be reused directly in the training script without needing to be nested in the fastpitch conf directory itself.
- Add code branches to FastPitch model to allow configuring new data loader for weighted sampling.
- Add `collate_fn` to featurizer class.
- Make log energy optional in FastPitch module so user can experiment with different normalization approaches.

# Usage
* This is an example command to train FastPitch on LJSpeech & VCTK with 0.2 and 0.8 sampling weight, respectively.

```bash
/home/rlangman/miniconda3/envs/nemo_dev/bin/python /home/rlangman/Code/NeMo/examples/tts/fastpitch.py
--config-path=/home/rlangman/Code/NeMo/examples/tts/conf/fastpitch
--config-dir=/home/rlangman/Code/NeMo/examples/tts/conf
--config-name=fastpitch_22050.yaml
n_speakers=110
speaker_path=/home/rlangman/Data/multi/ljspeech_vctk_speakers.json
max_epochs=1
weighted_sample_steps=10
phoneme_dict_path=/home/rlangman/Code/NeMo/scripts/tts_dataset_files/ipa_cmudict-0.7b_nv23.01.txt
heteronyms_path=/home/rlangman/Code/NeMo/scripts/tts_dataset_files/heteronyms-052722
feature_stats_path=/home/rlangman/Data/multi/ljspeech_vctk_stats.json
exp_manager.exp_dir=/home/rlangman/exps/ljspeech_vctk
+train_ds_meta.ljspeech._target_=nemo.collections.tts.data.text_to_speech_dataset.DatasetMeta
+train_ds_meta.ljspeech.manifest_path=/home/rlangman/Data/ljspeech/train_manifest.json
+train_ds_meta.ljspeech.audio_dir=/home/rlangman/Data/ljspeech/audio
+train_ds_meta.ljspeech.feature_dir=/home/rlangman/Data/ljspeech/features
+train_ds_meta.ljspeech.sample_weight=0.2
+val_ds_meta.ljspeech._target_=nemo.collections.tts.data.text_to_speech_dataset.DatasetMeta
+val_ds_meta.ljspeech.manifest_path=/home/rlangman/Data/ljspeech/val_manifest.json
+val_ds_meta.ljspeech.audio_dir=/home/rlangman/Data/ljspeech/audio
+val_ds_meta.ljspeech.feature_dir=/home/rlangman/Data/ljspeech/features
+train_ds_meta.vctk._target_=nemo.collections.tts.data.text_to_speech_dataset.DatasetMeta
+train_ds_meta.vctk.manifest_path=/home/rlangman/Data/vctk/train_manifest.json
+train_ds_meta.vctk.audio_dir=/home/rlangman/Data/vctk/audio
+train_ds_meta.vctk.feature_dir=/home/rlangman/Data/vctk/features
+train_ds_meta.vctk.sample_weight=0.8
+val_ds_meta.vctk._target_=nemo.collections.tts.data.text_to_speech_dataset.DatasetMeta
+val_ds_meta.vctk.manifest_path=/home/rlangman/Data/vctk/val_manifest.json
+val_ds_meta.vctk.audio_dir=/home/rlangman/Data/vctk/audio
+val_ds_meta.vctk.feature_dir=/home/rlangman/Data/vctk/features
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: nemo/collections/tts/parts/preprocessing/features.py [hunk 1]
```diff
@@ -333,3 +378,10 @@ def load(self, manifest_entry: dict, audio_dir: Path, feature_dir: Path) -> Dict
             feature_dir=feature_dir,
         )
         return feature_dict
+
+    def collate_fn(self, train_batch: List[Dict[str, Tensor]]) -> Dict[str, Tensor]:
+        feature_dict = {}
+        _collate_feature(feature_dict=feature_dict, feature_name=self.pitch_name, train_batch=train_batch)
+        _collate_feature(feature_dict=feature_dict, feature_name=self.voiced_mask_name, train_batch=train_batch)
+        _collate_feature(feature_dict=feature_dict, feature_name=self.voiced_prob_name, train_batch=train_batch)
```
Comment (redoctopus) [inline nemo/collections/tts/parts/preprocessing/features.py:386]: We're not guaranteed to have (or need) each of these features all the time. Should there be checks before each collate to make sure it's necessary?

File: nemo/collections/tts/parts/utils/tts_dataset_utils.py [hunk 2]
```diff
@@ -104,6 +104,31 @@ def general_padding(item, item_len, max_len, pad_value=0):
     return item
 
 
+def stack_tensors(tensors: List[torch.Tensor], max_lens: List[int], pad_value: float = 0.0) -> torch.Tensor:
```
Comment (redoctopus) [inline nemo/collections/tts/parts/utils/tts_dataset_utils.py:107]: Worth writing a few small tests for?

File: nemo/collections/tts/parts/utils/tts_dataset_utils.py [hunk 3]
```diff
@@ -150,3 +175,55 @@ def common_path(path1, path2):
...
+
+
+def get_weighted_sampler(
+    sample_weights: List[float], batch_size: int, num_steps: int
+) -> torch.utils.data.WeightedRandomSampler:
+    """
+    Create pytorch sampler for doing weighted random sampling.
+
+    Args:
+        sample_weights: List of sampling weights for all elements in the dataset.
```
Comment (redoctopus) [inline nemo/collections/tts/parts/utils/tts_dataset_utils.py:219]: Ordering is I guess the same as the manifest? What happens if some entries are filtered out?

Comment (redoctopus) [reply nemo/collections/tts/parts/utils/tts_dataset_utils.py:219]: Do we expect the user to set the sample weights? If so, we would want the correspondence between index and specific entry to be as clear as possible.

Mostly I'm worried about a situation where sample weights are set for the dataset in manifest order (since `self.data_samples` is constructed by reading through the manifest), but then some entry is skipped that the user didn't expect.

Comment (rlangman) [reply nemo/collections/tts/parts/utils/tts_dataset_utils.py:219]: In this data loader implementation the user only specifies how much weight to give to each training manifest (eg. in the usage example, 20% to LJSpeech and 80% to VCTK). Creating the item level sample weights (where are all items in the same manifest receive the same weight) is done internally.

We could also allow user to provide a "sample_weight" field in each row in the training manifest...but I could not think of any practical use case where someone would use that over the manifest-level weighting.

Unless your concern is that someone might use this utility function directly when defining their own data loader. In which case, it is a little error prone.

Comment (redoctopus) [reply nemo/collections/tts/parts/utils/tts_dataset_utils.py:219]: Ahhh. Alright, that makes sense to me 👍

**Review by redoctopus — APPROVED** [review]
LGTM

**Review by redoctopus — DISMISSED** [review]

**Review by redoctopus — APPROVED** [review]
